### DISCUSSION TITLE
App Store Receipt Verification Sandbox Fallback

### DISCUSSION THREAD

**Discussion by robert-nash** [imdhemy/laravel-in-app-purchases#28] [Q&A] [answered]
I am currently implementing this library but I've got a question about the behaviour of receipt validation/how this library is best used.

In the Apple documentation for verifyReceipt (https://developer.apple.com/documentation/appstorereceipts/verifyreceipt) it advises to verify the receipt first against the production endpoint and then to fallback to the sandbox if the status code 21007 is received.

The reasons they state for doing this seem to make sense to me, allowing you to use the same server for requests which might be coming from the sandbox and the production environment at the same time. For example, you might have live users sending production receipts at the same time as Apple reviewers are sending sandboxed receipts whilst testing new releases.

How do you deal with this when using this library? It seems that just using the 'appstore_sandbox' parameter would make it difficult to continue development on an app once in production.

I would be grateful on any thoughts on this.

Thanks your your work on this library!

**Comment by imdhemy** [comment, marked as answer]
Hi @robert-nash,
Thank you for starting this discussion! Would you mind adding a reply with answers to the below questions?

- Can you provide a staging server and use it to perform the testing?
- If No, How many Apple reviewers do you have on average?

## Update
This is handled in version [0.8.5](https://github.com/imdhemy/laravel-in-app-purchases/releases/tag/0.8.5). 🎉 🎉
Now, the **Verifier** uses the production URI by default. When it encounters the code `21007`, it resends the verification request to the sandbox URI.

**Reply by robert-nash** [reply]
Hi @imdhemy,

Thank you for your reply.

I could setup a staging server but I am not aware of anyway of giving Apple an alternative notification URL to use just for testing. I am new to IAP notifications though so if it was possible to do this that would be great to know.

I haven't gone through the review process for an app with IAP notifications before so I can't answer the second question unfortunately.

Thank you for your help

**Reply by imdhemy** [reply]
@robert-nash

Let me tell you how it works in our company's application. The mobile developers provide two bundles, one for production, and the other one is for internal testing. This allows us to use two notification URLs since they are actually two apps.

After careful consideration, using a config param would make it difficult as you mentioned. I'll work on enhancing this as soon as possible. 

Kindly, start an issue describing the problem and the expected output. I'll work on this issue as soon as possible.

**Reply by robert-nash** [reply]
Thank you, that makes sense and I see how that would get around much of the issue. Does this mean that Apple Reviewers for the production app are testing against a server using the production verifyReceipt endpoint?

I will open an issue for this later on. Thank you for all your existing work on this library and taking the time to make this enhancement.

**Reply by imdhemy** [reply]
Hi @robert-nash ,

This is handled in version [0.8.5](https://github.com/imdhemy/laravel-in-app-purchases/releases/tag/0.8.5). 🎉 🎉
Now, the **Verifier** uses the production URI by default. When it encounters the code `21007`, it resends the verification request to the sandbox URI.

**Reply by michavie** [reply]
Great work @imdhemy !

**Reply by imdhemy** [reply]
Thanks a lot @michavie !

 Your email answer was very helpful!

**Reply by robert-nash** [reply]
Thank you @imdhemy! Sorry I didn't get around to creating an issue. I will have a look at this new release.
